checker/tests/guieffect/Java8Lambdas.java

```diff
@@ -0,0 +1,176 @@
+            // needed.
+            arg.runOnUIThread(
+                    new IAsyncUITask() {
+                        final UIElement e2 = arg;
+
+                        @Override
+                        public void doStuff() { // should inherit UI effect
+                            func.executePolymorphic(e2); // should be okay
+                        }
+                    });
+        }
+    }
+
+    @PolyUIType
+    private static class PolymorphicLambdaRunner {
+        private final UIElement arg;
+
+        public PolymorphicLambdaRunner(UIElement arg) {
+            this.arg = arg;
+        }
+
+        @PolyUIEffect
+        public void doEither(@PolyUI PolymorphicFunctionalInterface<UIElement> func) {
+            func.executePolymorphic(this.arg);
+        }
+    }
+
+    public static void safeContextTestCases(UIElement elem) {
+        LambdaRunner runner = new LambdaRunner(elem);
+        runner.doSafe(e -> e.repaint());
+        //:: error: (call.invalid.ui)
+        runner.doSafe(e -> e.dangerous()); // Not allowed in doSafe
+        //:: error: (call.invalid.ui)
+        runner.doUI(e -> e.repaint()); // Not allowed in safe context
+        //:: error: (call.invalid.ui)
+        runner.doUI(e -> e.dangerous()); // Not allowed in safe context
+        runner.doEither(e -> e.repaint());
+        runner.doEither(e -> e.dangerous());
+        runner.doUISafely(e -> e.dangerous());
+        @AlwaysSafe PolymorphicLambdaRunner safePolymorphicLambdaRunner = new PolymorphicLambdaRunner(elem);
+        safePolymorphicLambdaRunner.doEither(e -> e.repaint());
+        // This next two are ok for this patch since the behavior is the same (no report) for lambdas as for annon
+        // classes. However, shouldn't this be (argument.type.incompatible) just because safePolymorphicLambdaRunner
+        // is not an @UI PolymorphicLambdaRunner ? Or, failing that (call.invalid.ui) since doEither is @PolyUIEffect ?
+        safePolymorphicLambdaRunner.doEither(e -> e.dangerous());
+        safePolymorphicLambdaRunner.doEither(
+                new @UI PolymorphicFunctionalInterface<UIElement>() {
+                    public void executePolymorphic(UIElement arg) {
+                        arg.dangerous();
+                    }
+                });
+        @UI PolymorphicLambdaRunner uiPolymorphicLambdaRunner = new @UI PolymorphicLambdaRunner(elem);
+        //:: error: (call.invalid.ui)
+        uiPolymorphicLambdaRunner.doEither(
+                e -> e.repaint()); // Safe at runtime, but not by the type system!
+        //:: error: (call.invalid.ui)
+        uiPolymorphicLambdaRunner.doEither(e -> e.dangerous());
+        PolymorphicFunctionalInterface<UIElement> func1 = e -> e.repaint();
+        // No error, why? //:: error: (assignment.type.incompatible)
+        PolymorphicFunctionalInterface<UIElement> func2 = e -> e.dangerous(); // Incompatible types!
```

Comment:
This test case should be failing, and I don't get why it isn't.

Reply:
I suspect this is related to bottom-up inference, and in hindsight I should have predicted this.

Bottom up inference lets us infer more lambdas' types correctly.  But, the checking that would have issued the ```assignment.type.incompatible``` occurs in the call to ```super.visitMethodInvocation(node, p)```, which is executed *before* you do the bottom-up inference of that call site.  So when the inherited method invocation checks are performed, the Checker Framework is still working only with the default ```@AlwaysSafe```.

Reply:
Doing a separate pass for inference before any checking occurs (which was the intent of my initial suggestion to put inference in the TreeAnnotator) would have the inferred ```@UI``` qualifiers in place before the subtyping checks run for passing actual parameters.  So when we find out what the proper solution is for separating inference in an earlier pass, that should fix these.

Reply:
I am not sure this is the true issue here.

I built a different version of this PR just now, which uses my own `SourceVisitor` to recursively annotate the lambdas (called from `GuiEffectVisitor.processClassTree` for every top level class). The results are exactly the same in terms of the test cases. Note that this version of the PR duplicates a fair amount of the checking code for assignments, method invocations and return after the recursive `super.visitX` calls.

Moving the inference to its own traversal does indeed make the code cleaner (requiring less duplicated checks), but there is a performance cost that I am not sure we want to pay to do it that way, particularly if the behavior is equivalent. Here is the diff between this PR and that alternative approach, in case it proves useful for the discussion: https://github.com/lazaroclapp/checker-framework/commit/2df04006495e6f6655538775c6080f656f19eabb

Reply:
I spent a while poking around the superclasses; as far as I can tell your ```getAnnotatedType``` override should be called.

This could still be a consequence of having multiple annotations on the type.  If I try to compile the following:

```java
import org.checkerframework.checker.guieffect.qual.*;

public class Blah {


    public @UI @AlwaysSafe Object o;

    public void m(@AlwaysSafe Object o) {
    }

    public void x() {
        m(o);
    }
}
```
I see the following complaints:
```
Blah.java:6: error: [type.invalid] [@AlwaysSafe, @UI] may not be applied to the type "@AlwaysSafe @UI Object"
    public @UI @AlwaysSafe Object o;
                                  ^
Blah.java:12: error: [type.invalid] [@AlwaysSafe, @UI] may not be applied to the type "@AlwaysSafe @UI Object"
        m(o);
          ^
2 errors
```
So blindly tacking on the ```@UI``` is probably breaking some sort of invariant --- @wmdietl would know for sure.

Have you tried explicitly removing the ```@AlwaysSafe``` from the underlying annotated type mirror in your override's lambda handling?

Reply:
This is absolutely right! Sorry about that. I just replaced the blind `addAnnotation` call with `replaceAnnotation` and now I seem to be getting *exactly* the same behavior as for anonymous classes. Not to say there aren't still a few issues with this patch (namely persisting the set of UI lambdas on the side), but it looks to me that it at least behaves correctly now. Thanks!